org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/CompilationUnitResolverDiscovery.java

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.jdt.core.dom;
+
+import java.util.Objects;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExtension;
+import org.eclipse.core.runtime.IExtensionPoint;
+import org.eclipse.core.runtime.ILog;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.internal.core.dom.ICompilationUnitResolver;
+
+class CompilationUnitResolverDiscovery {
+	private static final String SELECTED_SYSPROP = "ICompilationUnitResolver"; //$NON-NLS-1$
+	private static final String COMPILATION_UNIT_RESOLVER_EXTPOINT_ID = "compilationUnitResolver" ; //$NON-NLS-1$
+	private static boolean ERROR_LOGGED = false;
+
+	private static String lastId;
+	private static IConfigurationElement lastExtension;
+
+	static ICompilationUnitResolver getInstance() {
+		String id = System.getProperty(SELECTED_SYSPROP);
+		IConfigurationElement configElement = getConfigurationElement(id);
```

Comment:
As before, can we please avoid the code looking into the OSGI registry if system property is not set, to avoid that on every call we enter sync blocks and do extra work?

What I want is to make the "default" SDK case where property is **not** set as fast as it was before this change, and that is easily done by checking system property first.

Reply:
Please see the 1st lines of `getConfigurationElement(id)` method, they immediately return null when there is no setting, so extension registry is not even read.

Reply:
OK, this is a bit surprising, but I see it now.